### PR TITLE
mvcc: Introduce sample step for MVCC scanner

### PR DESCRIPTION
**Pull request by iosmanthus**
Signed-off-by: iosmanthus <myosmanthustree@gmail.com>

### What problem does this PR solve?

This pull request enables MVCC scanner to take a sample from keys/values while bypassing `sample_step - 1` keys between two continuous keys and ignore all locks (i.e. applies a read committed isolation). If `sample_step` is 0, sampling is disabled.

Problem Summary:

What's Changed:

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

* Introduce sample step for MVCC scanner

### CODE REVIEW

File: src/storage/mod.rs [hunk 1]
```diff
@@ -479,10 +480,17 @@ impl<E: Engine, L: LockManager, P: PdClient + 'static> Storage<E, L, P> {
                 {
                     let begin_instant = Instant::now_coarse();
 
+                    let isolation_level = if sample_step > 0 {
+                        limit *= sample_step;
+                        IsolationLevel::Rc
```
Comment (breeswish) [inline src/storage/mod.rs:485]: Why not pass RC directly?

Comment (iosmanthus) [reply src/storage/mod.rs:485]: Or else we have to modify the MVCC inner scanners and use some parameters to inform them to ignore locks which also introduces some state to them like an inner counter and an ignore-lock-flag.

Comment (breeswish) [reply src/storage/mod.rs:485]: Why not pass RC from TiDB? Please do not implement using HACK.

**Comment by coocood** [comment]
@breeswish @MyonKeminta PTAL

File: src/storage/mod.rs [hunk 2]
```diff
@@ -479,6 +480,10 @@ impl<E: Engine, L: LockManager, P: PdClient + 'static> Storage<E, L, P> {
                 {
                     let begin_instant = Instant::now_coarse();
 
+                    if sample_step > 0 {
+                        limit *= sample_step;
```
Comment (sticnarf) [inline src/storage/mod.rs:484]: Maybe `saturating_mul` is better in case of an overflow?

File: src/storage/txn/store.rs [hunk 3]
```diff
@@ -53,11 +53,18 @@ pub trait Scanner: Send {
     fn next(&mut self) -> Result<Option<(Key, Value)>>;
 
     /// Get the next [`KvPair`](KvPair)s up to `limit` if they exist.
```
Comment (sticnarf) [inline src/storage/txn/store.rs:55]: Please update the doc about sample_step

File: src/storage/mod.rs (hunk 2 above)
Comment (sticnarf) [inline src/storage/mod.rs:484]: Why need we change the limit? It seems to me that the `limit` used in `Scanner::scan` limits the final result size.

File: src/storage/txn/store.rs [hunk 4]
```diff
@@ -53,11 +53,18 @@ pub trait Scanner: Send {
...
+        let mut row_count = 0;
         let mut results = Vec::with_capacity(limit);
         while results.len() < limit {
             match self.next() {
                 Ok(Some((k, v))) => {
+                    if sample_step > 0 {
+                        if row_count % sample_step != 0 {
+                            continue;
+                        }
+                        row_count += 1;
```
Comment (sticnarf) [inline src/storage/txn/store.rs:66]: Should `row_count` be updated before the branch check?

Now, once `row_count % sample_step != 0` is true, `row_count` will never be changed and all other rows will be skipped.

**Comment by iosmanthus** [comment]
> Could you add a unit test for scanning with `sample_step`?

Done

File: src/storage/mod.rs (hunk 2 above)
Comment (iosmanthus) [reply src/storage/mod.rs:490]: Done

File: src/storage/txn/store.rs (hunk 3 above)
Comment (iosmanthus) [reply src/storage/txn/store.rs:55]: Done.

File: src/storage/mod.rs (hunk 2 above)
Comment (iosmanthus) [reply src/storage/mod.rs:490]: Removed.

**Comment by iosmanthus** [comment]
@sticnarf PTAL

**Pull request merged**